### ISSUE TITLE
[v4]  TextInput inside a bottomsheet have weird cursor bug. 

### ISSUE THREAD

**Issue by ilyaSuperGlue** [gorhom/react-native-scroll-bottom-sheet#867]
# Bug

BottomSheet have weird cursor bug on android, when i am trying to edit my textinput by moving the cursor.

## Environment info

| Library                         | Version |
| ------------------------------- | ------- |
| @gorhom/bottom-sheet            | 4.1.5  |
| react-native                    | 0.64.2  |
| react-native-reanimated         | 2.3.1  |
| react-native-gesture-handler    | 2.2.0  |

## Steps To Reproduce

1. create a bottom sheet component with TextInput
2.  trigger bottom sheet open
3.  type on text input then move the cursor to the start
4. start typing to edit text
5. cursor will jump to previous char making it unable to edit text

Video

https://user-images.githubusercontent.com/36694324/156755107-a02a41eb-64a0-4570-8491-04709d6d813a.mp4

Describe what you expected to happen:

1.  cursor should follow changes on text change

## Reproducible sample code

```javascript
import { View, Button } from 'react-native';
import React, { useMemo, useRef, useState } from 'react';
import {
  BottomSheetModal,
  BottomSheetModalProvider,
  BottomSheetBackdrop,
  BottomSheetTextInput,
} from '@gorhom/bottom-sheet';
import { BottomSheetModalMethods } from '@gorhom/bottom-sheet/lib/typescript/types';

const Index = () => {
  const [title, setTitle] = useState('');
  const snapPoints = useMemo(() => ['60%', '60%'], []);
  const itemRef = useRef<BottomSheetModalMethods>(null);
  return (
    <BottomSheetModalProvider>
      <View
        style={{
          flex: 1,
        }}>
        <Button onPress={() => itemRef.current?.present()} title="open modal" />
      </View>
      <BottomSheetModal
        snapPoints={snapPoints}
        index={1}
        backdropComponent={BottomSheetBackdrop}
        keyboardBehavior="interactive"
        keyboardBlurBehavior="restore"
        ref={itemRef}>
        <View
          style={{
            paddingHorizontal: 20,
          }}>
          <BottomSheetTextInput
            style={{
              borderBottomWidth: 1,
            }}
            placeholder="Title"
            value={title}
            onChangeText={(text) => setTitle(text)}
          />
        </View>
      </BottomSheetModal>
    </BottomSheetModalProvider>
  );
};

export default Index;

```

## React Native Info Result :

```
info Fetching system and libraries information...
System:
    OS: Windows 10 10.0.19044
    CPU: (16) x64 AMD Ryzen 7 4800H with Radeon Graphics
    Memory: 1.10 GB / 7.42 GB
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.17 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK:
      API Levels: 23, 28, 29, 30, 31
      Build Tools: 28.0.3, 29.0.2, 29.0.3, 30.0.2, 30.0.3, 31.0.0
      System Images: android-28 | Google APIs Intel x86 Atom_64
      Android NDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.19041.0
  IDEs:
    Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7784292
    Visual Studio: 16.11.32106.194 (Visual Studio Community 2019)
  Languages:
    Java: 1.8.0_41
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2
    react-native: 0.64.2 => 0.64.2
    react-native-windows: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

**Comment by pranavp10** [comment]
@hikaaam  the issue occurs due to the re-rendering of `BottomSheetModal`, `BottomSheetModalProvider`, `BottomSheetBackdrop,` on the change of the title field.

Here is the solution below👇🏻

```jsx
import { View, Button } from 'react-native';
import React, { useMemo, useRef, useState } from 'react';
import {
  BottomSheetModal,
  BottomSheetModalProvider,
  BottomSheetBackdrop,
  BottomSheetTextInput,
} from '@gorhom/bottom-sheet';
import { BottomSheetModalMethods } from '@gorhom/bottom-sheet/lib/typescript/types';

const InputComponent = () => {
  const [title, setTitle] = useState('');
  return (
    <BottomSheetTextInput
      style={{
        borderBottomWidth: 1,
      }}
      placeholder="Title"
      value={title}
      onChangeText={(text) => setTitle(text)}
    />
  )
}

const Index = () => {
  const snapPoints = useMemo(() => ['60%', '60%'], []);
  const itemRef = useRef<BottomSheetModalMethods>(null);
  return (
    <BottomSheetModalProvider>
      <View
        style={{
          flex: 1,
        }}>
        <Button onPress={() => itemRef.current?.present()} title="open modal" />
      </View>
      <BottomSheetModal
        snapPoints={snapPoints}
        index={1}
        backdropComponent={BottomSheetBackdrop}
        keyboardBehavior="interactive"
        keyboardBlurBehavior="restore"
        ref={itemRef}>
        <View
          style={{
            paddingHorizontal: 20,
          }}>
          <InputComponent />
        </View>
      </BottomSheetModal>
    </BottomSheetModalProvider>
  );
};

export default Index;
```

**Comment by romankhymych** [comment]
I'd had a flickering issue when using `TextInput` as a controlled component with `useState` and `multiline` prop inside BottomSheetModal (the same happens with BottomSheet).

`<TextInput value={text} onChangeText={setText} multiline />`

I found that only when TextInput is a controlled component and using `multiline` prop flickering during keystrokes happen.
Extracting children components outside of my component or using ref to prevent rendering on keystrokes did not fit my expectations. So I found the solution in the documentation of React Native about `defaultvalue` (https://reactnative.dev/docs/textinput#defaultvalue).

TL;DR
Just instead of `value` prop of `TextInfo` use `defaultvalue`

Full solution. Hope it helps.

```
import { FC, useState } from 'react';
import { TextInput } from 'react-native';

import { BottomSheet } from 'components/Elements/BottomSheet'; // BottomSheetModal
import FnButton from 'components/Elements/Button';

export const CustomComponent: FC = () => {
  const [text, setText] = useState('Default value');

  const handleSubmit = () => {
    setText('Test');
  };

  return (
    <BottomSheet>
      <>
        <TextInput defaultValue={text} onChangeText={setText} multiline />
        <FnButton onPress={handleSubmit} title="Submit" style={{ marginTop: 50 }} />
      </>
    </BottomSheet>
  );
};

```

**Comment by sondreluc** [comment]
Wow, thanks @romankhymych! I was experiencing this issue, and this solution fixed it!
I have a `handleComponent` that relies on the value of an text field inside the BottomSheetModal. I was therefor limited to keeping the  state for the text value outside of the sheet. 
~~What I don't understand tho, is why this works? The react native documentation for `defaultValue` does not explain it to me. 
The state value is updated as it should.~~ 
This seems to only work because the entire sheet content is being re-mounted on each change. Which is why replacing `value` with `defaultValue` doesn't brake the TextInput (`defaultValue` is only read by the input on mount, updating it afterwards should have no effect).
This re-mounting behavior seems like a bug to me. I'm going to investigate and as if I can find the cause of this in my own code, or if this has something to do with the internal code if `BottomSheetModal`
Why does this does not lead to the same flickers as with `value`, I don't understand.

**Comment by pehagg** [comment]
Stumbled into this issue today, spending an hour trying to figure out what happens. @romankhymych 's workaround did the trick, as confirmed above. A new issue should probably be reopened for the constant re-mounting issue, i.e. for fixing the root cause.

**Comment by sondreluc** [comment]
After further investigations, it turned out that the components where not re-mounting, as they first looked to be. 
My search for answers has lead me to this [issue](https://github.com/facebook/react-native/issues/35936) as well as other similar issues in the react native repo. The un-controled TextInput workaround was also mentioned in that issue. Not sure it actually is related, but at least it seems to be an issue with the TextInput component from RN, and not something to do with react-native-bottom-sheet.

**Comment by ilyaSuperGlue** [comment]
hi, i was the one that submit this issue,

for those that still having this issue, i remember fixing this issue by changing the `BottomSheetModal` to `BottomSheet`

seems like a bug on the `BottomSheetModal` component

**Comment by AmdjedSanero** [comment]
> I'd had a flickering issue when using `TextInput` as a controlled component with `useState` and `multiline` prop inside BottomSheetModal (the same happens with BottomSheet).
> 
> `<TextInput value={text} onChangeText={setText} multiline />`
> 
> I found that only when TextInput is a controlled component and using `multiline` prop flickering during keystrokes happen. Extracting children components outside of my component or using ref to prevent rendering on keystrokes did not fit my expectations. So I found the solution in the documentation of React Native about `defaultvalue` (https://reactnative.dev/docs/textinput#defaultvalue).
> 
> TL;DR Just instead of `value` prop of `TextInfo` use `defaultvalue`
> 
> Full solution. Hope it helps.
> 
> ```
> import { FC, useState } from 'react';
> import { TextInput } from 'react-native';
> 
> import { BottomSheet } from 'components/Elements/BottomSheet'; // BottomSheetModal
> import FnButton from 'components/Elements/Button';
> 
> export const CustomComponent: FC = () => {
>   const [text, setText] = useState('Default value');
> 
>   const handleSubmit = () => {
>     setText('Test');
>   };
> 
>   return (
>     <BottomSheet>
>       <>
>         <TextInput defaultValue={text} onChangeText={setText} multiline />
>         <FnButton onPress={handleSubmit} title="Submit" style={{ marginTop: 50 }} />
>       </>
>     </BottomSheet>
>   );
> };
> ```

**thank you**

**Comment by fansilva1991** [comment]
> I'd had a flickering issue when using `TextInput` as a controlled component with `useState` and `multiline` prop inside BottomSheetModal (the same happens with BottomSheet).
> 
> `<TextInput value={text} onChangeText={setText} multiline />`
> 
> I found that only when TextInput is a controlled component and using `multiline` prop flickering during keystrokes happen. Extracting children components outside of my component or using ref to prevent rendering on keystrokes did not fit my expectations. So I found the solution in the documentation of React Native about `defaultvalue` (https://reactnative.dev/docs/textinput#defaultvalue).
> 
> TL;DR Just instead of `value` prop of `TextInfo` use `defaultvalue`
> 
> Full solution. Hope it helps.
> 
> ```
> import { FC, useState } from 'react';
> import { TextInput } from 'react-native';
> 
> import { BottomSheet } from 'components/Elements/BottomSheet'; // BottomSheetModal
> import FnButton from 'components/Elements/Button';
> 
> export const CustomComponent: FC = () => {
>   const [text, setText] = useState('Default value');
> 
>   const handleSubmit = () => {
>     setText('Test');
>   };
> 
>   return (
>     <BottomSheet>
>       <>
>         <TextInput defaultValue={text} onChangeText={setText} multiline />
>         <FnButton onPress={handleSubmit} title="Submit" style={{ marginTop: 50 }} />
>       </>
>     </BottomSheet>
>   );
> };
> ```

Works perfectly! Thanks

**Comment by ViniDevBR** [comment]
This keep in 2026

**Comment by jikalmena** [comment]
The problem still persists. It's very frustrating.
